src/commissaire_service/clusterexec/__init__.py

```diff
@@ -0,0 +1,329 @@
+        model_instance.status = end_status
+
+        self.logger.info(
+            'Cluster {} final {} status: {}'.format(
+                cluster_name, command_name, model_instance.to_json()))
+
+        try:
+            params = {
+                'model_type_name': model_instance.__class__.__name__,
+                'model_json_data': model_instance.to_json()
+            }
+            self.request('storage.save', params=params)
+        except Exception as error:
+            self.logger.error(
+                'Unable to save final state for "{}" clusterexec '
+                'due to {}: {}'.format(cluster_name, type(error), error))
+
+    def on_upgrade(self, message, cluster_name):
+        """
+        Executes an upgrade command on hosts across a cluster.
+
+        :param message: A message instance
+        :type message: kombu.message.Message
+        :param cluster_name: The name of a cluster
+        :type cluster_name: str
+        """
+        self.logger.info(
+            'Received message: Upgrade cluster "{}"'.format(cluster_name))
+        command_args = ('upgrade',)
+        model_instance = ClusterUpgrade.new(
+            name=cluster_name,
+            status='in_process',
+            started_at=datetime.datetime.utcnow().isoformat(),
+            upgraded=[],
+            in_process=[]
+        )
+        self._execute(message, model_instance, command_args, 'upgraded')
+
+    def on_restart(self, message, cluster_name):
+        """
+        Executes a restart command on hosts across a cluster.
+
+        :param message: A message instance
+        :type message: kombu.message.Message
+        :param cluster_name: The name of a cluster
+        :type cluster_name: str
+        """
+        self.logger.info(
+            'Received message: Restart cluster "{}"'.format(cluster_name))
+        command_args = ('restart',)
+        model_instance = ClusterRestart.new(
+            name=cluster_name,
+            status='in_process',
+            started_at=datetime.datetime.utcnow().isoformat(),
+            restarted=[],
+            in_process=[]
+        )
+        self._execute(message, model_instance, command_args, 'restarted')
+
+    def on_deploy(self, message, cluster_name, version):
+        """
+        Executes a deploy command on atomic hosts across a cluster.
+
+        :param message: A message instance
+        :type message: kombu.message.Message
+        :param cluster_name: The name of a cluster
+        :type cluster_name: str
+        :param version: The tree image version to deploy
+        :type version: str
+        """
+        self.logger.info(
+            'Received message: Deploy version "{}" on cluster "{}"'.format(
+                version, cluster_name))
+        command_args = ('deploy', version)
+        model_instance = ClusterDeploy.new(
+            name=cluster_name,
+            status='in_process',
+            started_at=datetime.datetime.utcnow().isoformat(),
```

Comment:
nit: ``commissaire.util.date.formatted_dt`` could be used here since it's been merged.